sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -258,11 +253,18 @@ case class Substring(str: Expression, pos: Expression, len: Expression) extends
       null
     } else {
       val start = po.asInstanceOf[Int]
-      val length = ln.asInstanceOf[Int] 
-
+      val length = ln.asInstanceOf[Int]
       string match {
-        case ba: Array[Byte] => slice(ba, start, length)
-        case other => slice(other.toString, start, length)
+        case ba: Array[Byte] =>
+          val (st, end) = slicePos(start, length, () => ba.length)
+          ba.slice(st, end)
+        case s: UTF8String =>
+          val (st, end) = slicePos(start, length, () => s.length)
+          s.slice(st, end)
+        case other =>
```

Comment:
Yeah, we probably need to add a rule for this in `FunctionArgumentConversion` to explicitly cast.  I assume this is a hive test?  I know we are pretty sloppy about this right now, but I think it would be best to start tightening up all of these cases as you find them.